core/arch/arm/kernel/abort.c

```diff
@@ -63,41 +88,113 @@ static void __print_stack_unwind(struct abort_info *ai)
 	state.registers[9] = ai->regs->r9;
 	state.registers[10] = ai->regs->r10;
 	state.registers[11] = ai->regs->r11;
+
 	state.registers[13] = read_mode_sp(ai->regs->spsr & CPSR_MODE_MASK);
 	state.registers[14] = read_mode_lr(ai->regs->spsr & CPSR_MODE_MASK);
 	state.registers[15] = ai->pc;
 
 	do {
 		EMSG_RAW(" pc 0x%08x", state.registers[15]);
-	} while (unwind_stack(&state));
+	} while (unwind_stack_arm32(&state, exidx, exidx_sz));
 }
-#endif /*ARM32*/
+#else /* ARM32 */
+
+static uint32_t r32(uint64_t x)
+{
+	return x & 0xFFFFFFFF;
+}
+
+static void __print_stack_unwind_arm32(struct abort_info *ai __unused)
+{
+	struct unwind_state_arm32 state;
+	struct tee_ta_session *s;
+	struct user_ta_ctx *utc;
+	uaddr_t exidx;
+	size_t exidx_sz;
+
+	/* 64-bit kernel, hence 32-bit unwind must be for user mode */
+	assert(abort_is_user_exception(ai));
 
+	if (tee_ta_get_current_session(&s) != TEE_SUCCESS)
+		panic();
+
+	EMSG_RAW("Call stack:");
+	utc = to_user_ta_ctx(s->ctx);
+	exidx = utc->exidx_start;
+	exidx_sz = utc->exidx_size;
+	if (!exidx || !exidx_sz) {
+		/* Can't unwind, the PC is as much as we can show */
+		EMSG_RAW(" pc 0x%08x", ai->pc);
+		return;
+	}
+	exidx += utc->load_addr;
+	memset(&state, 0, sizeof(state));
+	state.registers[0] = r32(ai->regs->x0);
```

Comment:
Direct assignment should be OK, if you'd like to be very clear that we're truncating an `uint64_t` into an `uint32_t` a plain cast would be enough.